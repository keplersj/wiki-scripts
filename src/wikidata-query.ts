import got from "got";
import { stringify } from "query-string";

export async function queryWikidata(query: string) {
  return got(
    `https://query.wikidata.org/sparql?${stringify({
      query,
    })}`,
    { headers: { Accept: "application/sparql-results+json" } }
  ).json();
}
