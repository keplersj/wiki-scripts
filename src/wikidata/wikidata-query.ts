import got from "got";
import { stringify } from "query-string";

export interface QueryVariables {
  [key: string]: any;
}

export interface QueryResult<T extends QueryVariables> {
  head: {
    vars: keyof T;
  };
  results: {
    bindings: T[];
  };
}

export async function queryWikidata<T extends QueryVariables>(
  query: string
): Promise<QueryResult<T>> {
  return got(
    `https://query.wikidata.org/sparql?${stringify({
      query,
    })}`,
    { headers: { Accept: "application/sparql-results+json" } }
  ).json();
}

interface ItemsWithPropertyOptions {
  limit?: number;
}

const defaultItemsWithPropertyOptions: ItemsWithPropertyOptions = {};

export interface ItemsWithPropertyVariables {
  item: { value: string };
  itemLabel: { value: string };
  propertyValue: { value: string };
}

export async function itemsWithProperty(
  property: string,
  options: Partial<ItemsWithPropertyOptions> = defaultItemsWithPropertyOptions
) {
  const queryString = `
  SELECT DISTINCT ?item ?itemLabel ?propertyValue WHERE {
    ?item wdt:${property} ?propertyValue.
    SERVICE wikibase:label { bd:serviceParam wikibase:language "[AUTO_LANGUAGE]". }
  }
  
  ${options.limit ? `LIMIT ${options.limit}` : ""}
  `;

  const queryRes: QueryResult<ItemsWithPropertyVariables> = await queryWikidata(
    queryString
  );

  return queryRes;
}
