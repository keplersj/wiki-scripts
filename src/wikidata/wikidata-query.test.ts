const mockedGot = jest.fn();
mockedGot.mockReturnValue({
  json: () => Promise.resolve({ dataAttr1: "val11111" }),
} as any);

jest.mock("got", () => mockedGot);

beforeEach(() => {
  mockedGot.mockClear();
});

import { itemsWithProperty, queryWikidata } from "./wikidata-query";

describe("#queryWikidata", () => {
  it("calls the Wikidata query service as expected", async () => {
    const result = await queryWikidata(`
    SELECT DISTINCT ?item ?itemLabel ?npmPackageName WHERE {
      ?item wdt:P8262 ?npmPackageName.
      SERVICE wikibase:label { bd:serviceParam wikibase:language "[AUTO_LANGUAGE]". }
    }
    
    LIMIT 10
    `);

    expect(result).toMatchSnapshot();
    expect(mockedGot.mock.calls).toMatchSnapshot();
  });
});

describe("#itemsWithProperty", () => {
  it("calls the Wikidata query as expected", async () => {
    const result = await itemsWithProperty("P1234");

    expect(result).toMatchSnapshot();
    expect(mockedGot.mock.calls).toMatchSnapshot();
  });

  it("calls the Wikidata query as expected, with limit specified", async () => {
    const result = await itemsWithProperty("P1234", { limit: 1000 });

    expect(result).toMatchSnapshot();
    expect(mockedGot.mock.calls).toMatchSnapshot();
  });
});
