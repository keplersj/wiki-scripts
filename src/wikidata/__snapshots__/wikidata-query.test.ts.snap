// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#itemsWithProperty calls the Wikidata query as expected 1`] = `
Object {
  "dataAttr1": "val11111",
}
`;

exports[`#itemsWithProperty calls the Wikidata query as expected 2`] = `
Array [
  Array [
    "https://query.wikidata.org/sparql?query=%0A%20%20SELECT%20DISTINCT%20%3Fitem%20%3FitemLabel%20%3FpropertyValue%20WHERE%20%7B%0A%20%20%20%20%3Fitem%20wdt%3AP1234%20%3FpropertyValue.%0A%20%20%20%20SERVICE%20wikibase%3Alabel%20%7B%20bd%3AserviceParam%20wikibase%3Alanguage%20%22%5BAUTO_LANGUAGE%5D%22.%20%7D%0A%20%20%7D%0A%20%20%0A%20%20LIMIT%2010%0A%20%20",
    Object {
      "headers": Object {
        "Accept": "application/sparql-results+json",
      },
    },
  ],
]
`;

exports[`#itemsWithProperty calls the Wikidata query as expected, with limit specified 1`] = `
Object {
  "dataAttr1": "val11111",
}
`;

exports[`#itemsWithProperty calls the Wikidata query as expected, with limit specified 2`] = `
Array [
  Array [
    "https://query.wikidata.org/sparql?query=%0A%20%20SELECT%20DISTINCT%20%3Fitem%20%3FitemLabel%20%3FpropertyValue%20WHERE%20%7B%0A%20%20%20%20%3Fitem%20wdt%3AP1234%20%3FpropertyValue.%0A%20%20%20%20SERVICE%20wikibase%3Alabel%20%7B%20bd%3AserviceParam%20wikibase%3Alanguage%20%22%5BAUTO_LANGUAGE%5D%22.%20%7D%0A%20%20%7D%0A%20%20%0A%20%20LIMIT%201000%0A%20%20",
    Object {
      "headers": Object {
        "Accept": "application/sparql-results+json",
      },
    },
  ],
]
`;

exports[`#queryWikidata calls the Wikidata query service as expected 1`] = `
Object {
  "dataAttr1": "val11111",
}
`;

exports[`#queryWikidata calls the Wikidata query service as expected 2`] = `
Array [
  Array [
    "https://query.wikidata.org/sparql?query=%0A%20%20%20%20SELECT%20DISTINCT%20%3Fitem%20%3FitemLabel%20%3FnpmPackageName%20WHERE%20%7B%0A%20%20%20%20%20%20%3Fitem%20wdt%3AP8262%20%3FnpmPackageName.%0A%20%20%20%20%20%20SERVICE%20wikibase%3Alabel%20%7B%20bd%3AserviceParam%20wikibase%3Alanguage%20%22%5BAUTO_LANGUAGE%5D%22.%20%7D%0A%20%20%20%20%7D%0A%20%20%20%20%0A%20%20%20%20LIMIT%2010%0A%20%20%20%20",
    Object {
      "headers": Object {
        "Accept": "application/sparql-results+json",
      },
    },
  ],
]
`;
